questionpy/_ui.py

```diff
@@ -1,44 +1,79 @@
 from typing import TYPE_CHECKING
 
 import jinja2
 
-from questionpy_common.environment import Package, get_qpy_environment
+from questionpy_common.environment import Package, PackageNamespaceAndShortName, get_qpy_environment
+from questionpy_sdk.constants import TEMPLATES_DIR
 
 if TYPE_CHECKING:
     from questionpy import Attempt, Question
 
 
-def _loader_for_package(package: Package) -> jinja2.BaseLoader | None:
-    pkg_name = f"{package.manifest.namespace}.{package.manifest.short_name}"
-    if not (importlib.resources.files(pkg_name) / "templates").is_dir():
-        # The package has no "templates" directory, which would cause PackageLoader to raise an unhelpful ValueError.
+class _TraversableTemplateLoader(jinja2.BaseLoader):
+    """In contrast to the :class:`jinja2.FileSystemLoader` this does not support the auto-reload feature of jinja2."""
+
+    def __init__(self, traversable: Traversable):
+        self.traversable = traversable
+
+    def get_source(self, environment: jinja2.Environment, template: str) -> tuple[str, str, Callable[[], bool] | None]:
+        source_path = self.traversable.joinpath(template)
+        try:
+            source = source_path.read_text("utf-8")
+        except FileNotFoundError as e:
+            raise jinja2.TemplateNotFound(template) from e
+        else:
+            return source, template, None
+
+
+def _get_loader(package: Package) -> jinja2.BaseLoader | None:
+    templates_folder = package.get_path(f"{TEMPLATES_DIR}/")
+    if not templates_folder.is_dir():
         return None
 
-    # TODO: This looks for templates in python/<namespace>/<short_name>/templates, we might want to support a different
-    #  directory, such as resources/templates.
-    return jinja2.PackageLoader(pkg_name)
+    if os.path.exists(str(templates_folder)):
+        return jinja2.FileSystemLoader(str(templates_folder))
+    return _TraversableTemplateLoader(templates_folder)
 
 
 def create_jinja2_environment(attempt: "Attempt", question: "Question") -> jinja2.Environment:
     """Creates a Jinja2 environment with sensible default configuration.
 
     - Library templates are accessible under the prefix ``qpy/``.
-    - Package templates are accessible under the prefix ``<namespace>.<short_name>/``.
+    - Package templates are accessible under the prefix ``@<namespace>.<short_name>/``.
+    - The prefix is optional when accessing templates of the current package.
     - The QPy environment, attempt, question and question type are available as globals.
     """
     qpy_env = get_qpy_environment()
 
     loader_mapping = {}
     for package in qpy_env.packages.values():
-        loader = _loader_for_package(package)
+        loader = _get_loader(package)
         if loader:
-            loader_mapping[f"{package.manifest.namespace}.{package.manifest.short_name}"] = loader
+            loader_mapping[f"@{package.manifest.namespace}.{package.manifest.short_name}"] = loader
```

Comment:
Wenn es ein Package namens `foo.bar` und ein Package namens `foo.bar_baz` gibt, wird das Template `@foo.bar_baz/formulation.xhtml.j2` fälschlicherweise im Paket `foo.bar` gesucht. Könnte der Schrägstrich mit in den Präfix aufgenommen werden?

```suggestion
            loader_mapping[f"@{package.manifest.namespace}.{package.manifest.short_name}/"] = loader
```

Reply:
> Wenn es ein Package namens `foo.bar` und ein Package namens `foo.bar_baz` gibt, wird das Template `@foo.bar_baz/formulation.xhtml.j2` fälschlicherweise im Paket `foo.bar` gesucht.

Bist du dir sicher? Das scheint mir nämlich nicht der Fall zu sein, siehe [PrefixLoader.get_loader](https://github.com/pallets/jinja/blob/ada0a9a6fc265128b46949b5144d2eaa55e6df2c/src/jinja2/loaders.py#L514-L520).

Reply:
My bad, ich habe mir den PrefixLoader nicht genau angeschaut.